jiant/tasks/tasks.py

```diff
@@ -3731,3 +3732,113 @@ def get_metrics(self, reset=False):
         """Get metrics specific to the task"""
         acc = self.scorer1.get_metric(reset)
         return {"accuracy": acc}
+
+
+@register_task("wikipedia_corpus_sop", rel_path="wikipedia_corpus_small")
+class SentenceOrderTask(PairClassificationTask):
+    """ Task class for Sentence Order Prediction (SOP). See the ALBERT paper for details on SOP:
+        https://arxiv.org/abs/1909.11942.
+        We are currently using an unpreprocessed version of the Wikipedia corpus
+        (more specifically, the Wikidump data) that consists of 5% of the data. You can generate
+        the data by following the instructions from jiant/scripts/mlm.
+        One thing to note about our SOP ALBERT implementationn is that we do not load the pretrained
+        weights for the SOP head beacuse they are unavailable in Huggingface. We only use the
+        pretrained weights of the linear layer from ALBERT that creates the pooled output used in SOP.
+    """
+
+    def __init__(self, path, max_seq_len, name, **kw):
+        super(SentenceOrderTask, self).__init__(name, n_classes=2, **kw)
+        self.path = path
+        self.max_seq_len = max_seq_len
+        self.train_data_text = None
+        self.val_data_text = None
+        self.test_data_text = None
+        self.files_by_split = {
+            "train": os.path.join(path, "train.txt"),
+            "val": os.path.join(path, "valid.txt"),
+            "test": os.path.join(path, "test.txt"),
+        }
+        self._label_namespace = self.name + "_labels"
+
+    def get_data_iter(self, path):
+        """Loading data file and tokenizing the text. We override the
+        this function and all functions that call this function because
+        the step of reading in the data for SOP is different than other
+        PairClassificationTasks.
+
+        Args:
+            path: (str) data file path
+        """
+        f = open(path, "r")
+        #  The dataset comes with one sentence per line, thus we split by
+        #  line here.
+        text = [line.strip() for line in f]
```

Comment:
This looks like it loads the whole files into memory. Since we're ultimately yielding, it would be better to use a generator here as well. E.g.:

```python
prev_sent = None
this_sent = None
for line in f:
    prev_sent = this_sent
    this_sent = line.strip()
    if prev_sent is None:
        continue
    is_right_order = random.uniform(0, 1) > 0.5
    if is_right_order:
        sent_a, sent_b = prev_sent, this_sent
    else:
        sent_a, sent_b = this_sent, prev_sent
    sent_a_processed = ...
    sent_b_processed = ...
    yield ...
```